src/aviary/env.py

```diff
@@ -234,11 +237,32 @@ async def _exec_tool_call(tool_call: ToolCall) -> ToolResponseMessage:
                 s_content = json.dumps(content)
             return ToolResponseMessage.from_call(tool_call, content=s_content)
 
+        invalid_msgs = []
+        valid_action = message
+        call_ordering = [t.id for t in message.tool_calls]
+        if handle_invalid_tool_calls:
+            valid_action, invalid_action = self.filter_invalid_tool_calls(message)
+            invalid_msgs = [
+                ToolResponseMessage.from_call(
+                    tool_call, content=f"Invalid tool call: {tool_call.function.name}"
+                )
+                for tool_call in invalid_action.tool_calls
+            ]
+
         if not ordered:
-            return await asyncio.gather(
-                *(_exec_tool_call(tool_call) for tool_call in message.tool_calls)
+            result = invalid_msgs + (
+                await asyncio.gather(
+                    *(
+                        _exec_tool_call(tool_call)
+                        for tool_call in valid_action.tool_calls
+                    )
+                )
             )
-        return [await _exec_tool_call(tool_call) for tool_call in message.tool_calls]
+            return sorted(result, key=lambda x: call_ordering.index(x.tool_call_id))
+
+        return invalid_msgs + [
+            await _exec_tool_call(tool_call) for tool_call in valid_action.tool_calls
+        ]
```

Comment:
Are you sure this is the same order? It seems like it will always be invalid tool calls lumped up front, and valid tool calls lumped in the back, which may not be their original order

Reply:
yea - that was a fuck up. Added a unit test and fixed